Assertion `srq->tail == srq_info.tail' in */test_ucp_ep_force_disconnect.test/2

http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5001/label=hpc-test-node,worker=2/console
```
13:22:03 [ RUN      ] shm/test_ucp_ep_force_disconnect.test/2
13:22:03 [hpc-test-node:16818:0] ib_mlx5_log.c:109  FATAL: Error on QP 0x24dfb wqe[467]: Remote side aborted (synd 0x22 vend 0x99) opcode RECV
13:22:03 
13:22:03 /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c: [ uct_ib_mlx5_completion_with_err() ]
13:22:03       ...
13:22:03       104     }
13:22:03       105 
13:22:03       106     ucs_log(log_level, "Error on QP 0x%x wqe[%d]: %s (synd 0x%x vend 0x%x) opcode %s",
13:22:03 ==>   107             qp_num, wqe_counter, info, ecqe->syndrome, ecqe->vendor_err_synd,
13:22:03       108             uct_ib_mlx5_cqe_err_opcode(ecqe));
13:22:03       109 }
13:22:03       110 
13:22:03 
13:22:04 ==== backtrace ====
13:22:04  0 0x000000000002c0ca uct_ib_mlx5_completion_with_err()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5_log.c:107
13:22:04  1 0x000000000002cfc0 uct_ib_mlx5_check_completion()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5.c:258
13:22:04  2 0x0000000000048b44 uct_ib_mlx5_poll_cq()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/mlx5/ib_mlx5.inl:33
13:22:04  3 0x000000000003c7f4 uct_rc_mlx5_ep_t_cleanup()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/rc/accel/rc_mlx5_ep.c:461
13:22:04  4 0x00000000000571d6 ucs_class_call_cleanup_chain()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucs/type/class.c:50
13:22:04  5 0x0000000000046f80 uct_rc_mlx5_ep_t_delete()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/uct/ib/rc/accel/rc_mlx5_ep.c:483
13:22:04  6 0x0000000000015b81 ucp_ep_destroy_internal()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_ep.c:348
13:22:04  7 0x0000000000020999 ucp_worker_destroy_eps()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_worker.c:1132
13:22:04  8 0x0000000000020999 ucp_worker_destroy()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../src/ucp/core/ucp_worker.c:1140
13:22:04  9 0x000000000062c421 ucs::handle<ucp_worker*, void*>::release()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:356
13:22:04 10 0x000000000062c421 ucs::handle<ucp_worker*, void*>::reset()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:295
13:22:04 11 0x000000000062c421 ~handle()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:290
13:22:04 12 0x000000000062c421 ~pair()  /usr/include/c++/4.8.2/bits/stl_pair.h:96
13:22:04 13 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >()  /usr/include/c++/4.8.2/bits/stl_construct.h:93
13:22:04 14 0x000000000062c421 __destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*>()  /usr/include/c++/4.8.2/bits/stl_construct.h:103
13:22:04 15 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*>()  /usr/include/c++/4.8.2/bits/stl_construct.h:126
13:22:04 16 0x000000000062c421 _Destroy<std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > >*, std::pair<ucs::handle<ucp_worker*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >()  /usr/include/c++/4.8.2/bits/stl_construct.h:151
13:22:04 17 0x000000000062c421 std::vector<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >, std::allocator<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > > > >::_M_erase_at_end()  /usr/include/c++/4.8.2/bits/stl_vector.h:1352
13:22:04 18 0x000000000062c421 std::vector<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > >, std::allocator<std::pair<ucs::handle<ucp_worker*, void*>, std::vector<ucs::handle<ucp_ep*, ucp_test_base::entity*>, std::allocator<ucs::handle<ucp_ep*, ucp_test_base::entity*> > > > > >::clear()  /usr/include/c++/4.8.2/bits/stl_vector.h:1126
13:22:04 19 0x000000000062c421 ucp_test_base::entity::cleanup()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/ucp_test.cc:463
13:22:04 20 0x000000000062c421 ucp_test_base::entity::~entity()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/ucp_test.cc:338
13:22:04 21 0x000000000057177e ucs::ptr_vector_base<ucp_test_base::entity>::release()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:223
13:22:04 22 0x000000000057177e ucs::ptr_vector<ucp_test_base::entity>::remove()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test_helpers.h:246
13:22:04 23 0x0000000000565379 test_ucp_peer_failure::fail_receiver()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:97
13:22:04 24 0x0000000000565379 ucs::entities_storage<ucp_test_base::entity>::sender()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.h:134
13:22:04 25 0x0000000000565379 test_ucp_peer_failure::test_force_close()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:197
13:22:04 26 0x000000000046e6c6 ucs::test_base::run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.cc:251
13:22:04 27 0x000000000046e6c6 ucs::test_base::TestBodyProxy()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/test.cc:277
13:22:04 28 0x00000000004663c3 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3562
13:22:04 29 0x000000000045a83d testing::Test::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3635
13:22:04 30 0x000000000045a90c testing::TestInfo::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3812
13:22:04 31 0x000000000045aa6f testing::TestCase::Run()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:3930
13:22:04 32 0x000000000045f407 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:5802
13:22:04 33 0x000000000045f70b testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest-all.cc:5719
13:22:04 34 0x000000000040f633 RUN_ALL_TESTS()  /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/worker/2/contrib/../test/gtest/common/gtest.h:20059
13:22:04 35 0x0000000000021b15 __libc_start_main()  ???:0
13:22:04 36 0x0000000000444fc9 _start()  ???:0
13:22:04 ===================
13:22:04 [1509535323.154023] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.196428] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.277014] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 [1509535323.321537] [hpc-test-node:16818:0]    ucp_context.c:392  UCX  INFO  No accelerator devices are available
13:22:04 Sending notification to devendar@mellanox.com
13:22:07 [hpc-test-node:16818:0] Process frozen...
```

happened again on shm_ib
http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6469/label=hpc-test-node2,worker=3/consoleFull#-853622487b816df86-9afc-46eb-984f-1a9a125d73bb
```
18:18:28 [ RUN      ] shm_ib/test_ucp_ep_force_disconnect.test/1
18:18:29 [hpc-test-node2:9211 :0:9211]     ib_mlx5.c:438  Assertion `srq->tail == srq_info.tail' failed: srq->tail=4095 srq_info.tail=460
18:18:29 
18:18:29 /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/uct/ib/mlx5/ib_mlx5.c: [ uct_ib_mlx5_srq_cleanup() ]
18:18:29       ...
18:18:29       434     status = uct_ib_mlx5_get_srq_info(verbs_srq, &srq_info);
18:18:29       435     ucs_assert_always(status == UCS_OK);
18:18:29       436     ucs_assertv_always(srq->tail == srq_info.tail, "srq->tail=%d srq_info.tail=%d",
18:18:29 ==>   437                        srq->tail, srq_info.tail);
18:18:29       438 }
18:18:29 
18:18:30 ==== backtrace ====
18:18:30  0 0x0000000000033edf uct_ib_mlx5_srq_cleanup()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/uct/ib/mlx5/ib_mlx5.c:437
18:18:30  1 0x0000000000033edf uct_ib_mlx5_get_srq_info()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/uct/ib/mlx5/ib_mlx5.c:111
18:18:30  2 0x0000000000033edf uct_ib_mlx5_srq_cleanup()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/uct/ib/mlx5/ib_mlx5.c:435
18:18:30  3 0x0000000000058ab6 ucs_class_call_cleanup_chain()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/ucs/type/class.c:50
18:18:30  4 0x00000000000526f0 uct_rc_mlx5_ep_t_delete()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/uct/ib/rc/accel/rc_mlx5_ep.c:608
18:18:30  5 0x00000000000159e1 ucp_ep_cleanup_lanes()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/ucp/core/ucp_ep.c:531
18:18:30  6 0x000000000002130a ucp_worker_destroy_eps()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../src/ucp/core/ucp_worker.c:1227
18:18:30  7 0x0000000000673689 ucs::handle<ucp_worker*, void*>::release()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/test_helpers.h:521
18:18:30  8 0x000000000067371b ucp_test_base::entity::~entity()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/ucp/ucp_test.cc:353
18:18:30  9 0x000000000059c49e ucs::ptr_vector_base<ucp_test_base::entity>::release()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/test_helpers.h:387
18:18:30 10 0x0000000000590208 test_ucp_peer_failure::test_force_close()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:232
18:18:30 11 0x0000000000479266 ucs::test_base::run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/test.cc:270
18:18:30 12 0x000000000046d9c3 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:3562
18:18:30 13 0x0000000000461e3d testing::Test::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:3635
18:18:30 14 0x0000000000461f0c testing::TestInfo::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:3812
18:18:30 15 0x000000000046206f testing::TestCase::Run()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:3930
18:18:30 16 0x0000000000466a07 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:5802
18:18:30 17 0x0000000000466d0b testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest-all.cc:5719
18:18:30 18 0x000000000040fd33 main()  /scrap/jenkins/workspace/hpc-ucx-pr/label/hpc-test-node2/worker/3/contrib/../test/gtest/common/gtest.h:20059
18:18:30 19 0x0000000000021c05 __libc_start_main()  ???:0
18:18:30 20 0x000000000044c5c0 _start()  ???:0
18:18:30 ===================
18:18:30 Sending notification to sergeyo@mellanox.com
18:18:32 [hpc-test-node2:9211 :0:9211] Process frozen...
20:47:44 
make: *** [test] Terminated
20:47:44 
Build step 'Execute shell' marked build as failure
```

--- a/config/m4/ib.m4
+++ b/config/m4/ib.m4
@@ -303,6 +303,8 @@ AS_IF([test "x$with_ib" == xyes],
                [])
            ])
 
+       AC_CHECK_DECLS([ibv_cmd_modify_qp],
+                      [], [], [[#include <infiniband/driver.h>]])
 
        mlnx_valg_libdir=$with_verbs/lib${libsuff}/mlnx_ofed/valgrind
        AC_MSG_NOTICE([Checking OFED valgrind libs $mlnx_valg_libdir])

--- a/src/uct/ib/rc/accel/rc_mlx5_ep.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_ep.c
@@ -6,6 +6,9 @@
 
 #include "rc_mlx5.h"
 
+#if HAVE_DECL_IBV_CMD_MODIFY_QP
+#include <infiniband/driver.h>
+#endif
 #include <uct/ib/mlx5/ib_mlx5_log.h>
 #include <ucs/arch/cpu.h>
 #include <ucs/sys/compiler.h>
@@ -571,6 +574,32 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, uct_iface_h tl_iface)
     return UCS_OK;
 }
 
+static void uct_rc_mlx5_ep_reset_qp(uct_rc_mlx5_ep_t *ep)
+{
+    uct_rc_txqp_t *txqp = &ep->super.txqp;
+
+    /* Make the HW generate CQEs for all in-progress SRQ receives from the QP,
+     * so we clean them all before ibv_modify_qp() can see them.
+     */
+#if HAVE_DECL_IBV_CMD_MODIFY_QP
+    struct ibv_qp_attr qp_attr;
+    struct ibv_modify_qp cmd;
+    int ret;
+
+    /* Bypass mlx5 driver, and go directly to command interface, to avoid
+     * cleaning the CQ in mlx5 driver
+     */
+    memset(&qp_attr, 0, sizeof(qp_attr));
+    qp_attr.qp_state = IBV_QPS_RESET;
+    ret = ibv_cmd_modify_qp(txqp->qp, &qp_attr, IBV_QP_STATE, &cmd, sizeof(cmd));
+    if (ret) {
+        ucs_warn("modify qp 0x%x to RESET failed: %m", txqp->qp->qp_num);
+    }
+#else
+    (void)uct_rc_modify_qp(txqp, IBV_QPS_ERR);
+#endif
+}
+
 static UCS_CLASS_CLEANUP_FUNC(uct_rc_mlx5_ep_t)
 {
     uct_rc_mlx5_iface_t *iface = ucs_derived_of(self->super.super.super.iface,
@@ -578,11 +607,7 @@ static UCS_CLASS_CLEANUP_FUNC(uct_rc_mlx5_ep_t)
 
     uct_ib_mlx5_txwq_cleanup(&self->tx.wq);
 
-    /* Modify QP to error to make HW generate CQEs for all in-progress SRQ
-     * receives from the QP, so we clean them all before ibv_modify_qp() can
-     * see them.
-     */
-    (void)uct_rc_modify_qp(&self->super.txqp, IBV_QPS_ERR);
+    uct_rc_mlx5_ep_reset_qp(self);
     uct_rc_mlx5_iface_commom_clean_srq(&iface->mlx5_common, &iface->super,
                                        self->qp_num);
 

--- a/src/uct/ib/rc/base/rc_ep.c
+++ b/src/uct/ib/rc/base/rc_ep.c
@@ -291,8 +291,8 @@ ucs_status_t uct_rc_modify_qp(uct_rc_txqp_t *txqp, enum ibv_qp_state state)
     memset(&qp_attr, 0, sizeof(qp_attr));
     qp_attr.qp_state = state;
     if (ibv_modify_qp(txqp->qp, &qp_attr, IBV_QP_STATE)) {
-        ucs_warn("modify qp 0x%x to state %d failed: %m", state,
-                 txqp->qp->qp_num);
+        ucs_warn("modify qp 0x%x to state %d failed: %m", txqp->qp->qp_num,
+                 state);
         return UCS_ERR_IO_ERROR;
     }
 

--- a/src/uct/ib/rc/base/rc_ep.c
+++ b/src/uct/ib/rc/base/rc_ep.c
@@ -291,7 +291,8 @@ ucs_status_t uct_rc_modify_qp(uct_rc_txqp_t *txqp, enum ibv_qp_state state)
     memset(&qp_attr, 0, sizeof(qp_attr));
     qp_attr.qp_state = state;
     if (ibv_modify_qp(txqp->qp, &qp_attr, IBV_QP_STATE)) {
-        ucs_warn("modify qp 0x%x to RESET failed: %m", txqp->qp->qp_num);
+        ucs_warn("modify qp 0x%x to state %d failed: %m", state,
+                 txqp->qp->qp_num);
         return UCS_ERR_IO_ERROR;
     }
 

